From 1a1b07cfd2c3eb3bf836979170e6ef38fbc0fea4 Mon Sep 17 00:00:00 2001
From: Evan Wang <xswang@marvell.com>
Date: Thu, 11 Jan 2018 17:37:29 +0800
Subject: [PATCH 2163/2241] arm64: dts: marvell: add vqmmc regulator on a7k DB
 for UHS

On A7K DB CP0, for some SD cards they can not be detected
or detected information wrong. The reason is that for UHS
card vqmmc should be 1.8V and without vqmmc regulator the
voltage is fixed to 3.3v and will fail to swicth to UHS
mode.

The patch added vqmmc regulator for A7K boards and updated
the 'armada7040-db-setup.txt' with the GPIO configuration.

Change-Id: I15fb14dde394daa078ee0ba30bed7640d0b627b7
Signed-off-by: Evan Wang <xswang@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/48843
Tested-by: iSoC Platform CI <ykjenk@marvell.com>
Reviewed-by: Omri Itach <omrii@marvell.com>
---
 Documentation/mvebu/a7k-a8k/armada7040-db-setup.txt |  4 ++++
 arch/arm64/boot/dts/marvell/armada-7040-db.dtsi     | 13 +++++++++++++
 2 files changed, 17 insertions(+)

diff --git a/Documentation/mvebu/a7k-a8k/armada7040-db-setup.txt b/Documentation/mvebu/a7k-a8k/armada7040-db-setup.txt
index eaa35eb..5ee640a 100644
--- a/Documentation/mvebu/a7k-a8k/armada7040-db-setup.txt
+++ b/Documentation/mvebu/a7k-a8k/armada7040-db-setup.txt
@@ -147,3 +147,7 @@ Network configuration
  eth0           | CP0   | eth0          | 0             | SGMII0 (1G)                   |
  eth1           | CP0   | eth1          | 2             | RGMII0                        |
  ----------------------------------------------------------------------------------------
+
+I2C IO expander0 GPIO pin configuration
+--------------------------------------
+PIN[15] - GPIO regulator of cp0-vccq-sd0
diff --git a/arch/arm64/boot/dts/marvell/armada-7040-db.dtsi b/arch/arm64/boot/dts/marvell/armada-7040-db.dtsi
index 62c49d3..6b13174 100644
--- a/arch/arm64/boot/dts/marvell/armada-7040-db.dtsi
+++ b/arch/arm64/boot/dts/marvell/armada-7040-db.dtsi
@@ -109,6 +109,7 @@
 				pinctrl-names = "default";
 				pinctrl-0 = <&cp0_sdhci_pins>;
 				bus-width = <4>;
+				vqmmc-supply = <&cp0_vccq_sd0_reg>;
 			};
 			spi@700680 {
 				spi-flash@0 {
@@ -215,4 +216,16 @@
 		gpio = <&cp0_gpio0 12 GPIO_ACTIVE_HIGH>;
 		status = "disabled";
 	};
+	cp0_vccq_sd0_reg: cp0_vccq_sd0 {
+		compatible = "regulator-gpio";
+		regulator-name = "cp0-vccq-sd0";
+		regulator-min-microvolt = <1800000>;
+		regulator-max-microvolt = <3300000>;
+		regulator-boot-on;
+		gpios = <&expander0 15 GPIO_ACTIVE_HIGH>;
+		gpios-states = <0>;
+		states = <1800000 0x1
+			  3300000 0x0>;
+		enable-active-high;
+	};
 };
-- 
2.7.4

